File: golang/tests/pkg/integration/env_integration_test.go
package integration

import (
	"os"
	"testing"

	"github.com/aliyun/wuying-agentbay-sdk/golang/pkg/agentbay"
)

const envTestImage = "imgc-0ab5takivb1ke11hu"

func TestEnvSetAndGet(t *testing.T) {
	apiKey := os.Getenv("AGENTBAY_API_KEY")
	if apiKey == "" {
		t.Skip("AGENTBAY_API_KEY not set")
	}

	ab, err := agentbay.NewAgentBay(apiKey, nil)
	if err != nil {
		t.Fatalf("Failed to create AgentBay: %v", err)
	}

	result, err := ab.Create(agentbay.NewCreateSessionParams().WithImageId(envTestImage))
	if err != nil {
		t.Fatalf("Failed to create session: %v", err)
	}
	session := result.Session
	defer func() {
		_, _ = session.Delete()
	}()

	// Test Set
	setResult, err := session.Env.Set(map[string]string{
		"TEST_KEY": "test_value",
		"SDK_VER":  "1.0",
	})
	if err != nil {
		t.Fatalf("Env.Set error: %v", err)
	}
	if !setResult.Success {
		t.Fatalf("Env.Set failed: %s", setResult.ErrorMessage)
	}

	// Test Get specific keys
	getResult, err := session.Env.Get("TEST_KEY", "SDK_VER")
	if err != nil {
		t.Fatalf("Env.Get error: %v", err)
	}
	if !getResult.Success {
		t.Fatalf("Env.Get failed: %s", getResult.ErrorMessage)
	}
	if getResult.Envs["TEST_KEY"] != "test_value" {
		t.Errorf("Expected TEST_KEY=test_value, got %s", getResult.Envs["TEST_KEY"])
	}
	if getResult.Envs["SDK_VER"] != "1.0" {
		t.Errorf("Expected SDK_VER=1.0, got %s", getResult.Envs["SDK_VER"])
	}

	// Test Get all
	allResult, err := session.Env.Get()
	if err != nil {
		t.Fatalf("Env.Get() all error: %v", err)
	}
	if !allResult.Success {
		t.Fatalf("Env.Get() all failed: %s", allResult.ErrorMessage)
	}
	if allResult.Envs["TEST_KEY"] != "test_value" {
		t.Errorf("Expected TEST_KEY in all envs, got %s", allResult.Envs["TEST_KEY"])
	}

	// Test overwrite
	_, err = session.Env.Set(map[string]string{"TEST_KEY": "updated"})
	if err != nil {
		t.Fatalf("Env.Set overwrite error: %v", err)
	}
	updated, _ := session.Env.Get("TEST_KEY")
	if updated.Envs["TEST_KEY"] != "updated" {
		t.Errorf("Expected TEST_KEY=updated after overwrite, got %s", updated.Envs["TEST_KEY"])
	}

	// Test visible in shell
	_, _ = session.Env.Set(map[string]string{"SHELL_VIS": "hello_from_env"})
	cmdResult, err := session.Command.ExecuteCommand("echo $SHELL_VIS", 5000)
	if err != nil {
		t.Fatalf("ExecuteCommand error: %v", err)
	}
	if !cmdResult.Success {
		t.Fatalf("Shell command failed: %s", cmdResult.ErrorMessage)
	}
	if cmdResult.Stdout == "" || cmdResult.Stdout != "hello_from_env\n" {
		t.Errorf("Expected shell to see SHELL_VIS, got stdout=%q", cmdResult.Stdout)
	}
}

func TestEnvSetEmptyReturnsError(t *testing.T) {
	apiKey := os.Getenv("AGENTBAY_API_KEY")
	if apiKey == "" {
		t.Skip("AGENTBAY_API_KEY not set")
	}

	ab, err := agentbay.NewAgentBay(apiKey, nil)
	if err != nil {
		t.Fatalf("Failed to create AgentBay: %v", err)
	}
	result, err := ab.Create(agentbay.NewCreateSessionParams().WithImageId(envTestImage))
	if err != nil {
		t.Fatalf("Failed to create session: %v", err)
	}
	session := result.Session
	defer func() {
		_, _ = session.Delete()
	}()

	_, err = session.Env.Set(map[string]string{})
	if err == nil {
		t.Error("Expected error when setting empty envs, got nil")
	}
}

File: java/agentbay/src/integration-test/java/com/aliyun/agentbay/env/EnvIntegrationTest.java
package com.aliyun.agentbay.env;

import com.aliyun.agentbay.AgentBay;
import com.aliyun.agentbay.exception.AgentBayException;
import com.aliyun.agentbay.model.BoolResult;
import com.aliyun.agentbay.model.CommandResult;
import com.aliyun.agentbay.model.EnvResult;
import com.aliyun.agentbay.model.SessionResult;
import com.aliyun.agentbay.session.CreateSessionParams;
import com.aliyun.agentbay.session.Session;
import org.junit.jupiter.api.AfterEach;
import org.junit.jupiter.api.Assumptions;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;

import java.util.HashMap;
import java.util.Map;

import static org.junit.jupiter.api.Assertions.*;

public class EnvIntegrationTest {
    private static final String ENV_TEST_IMAGE = "imgc-0ab5takivb1ke11hu";
    private AgentBay agentBay;
    private Session session;

    @BeforeEach
    void setUp() throws AgentBayException {
        String apiKey = System.getenv("AGENTBAY_API_KEY");
        Assumptions.assumeTrue(apiKey != null && !apiKey.isEmpty(), "AGENTBAY_API_KEY not set");
        agentBay = new AgentBay(apiKey);
        CreateSessionParams params = new CreateSessionParams();
        params.setImageId(ENV_TEST_IMAGE);
        SessionResult sessionResult = agentBay.create(params);
        assertTrue(sessionResult.isSuccess(), "Failed to create session: " + sessionResult.getErrorMessage());
        session = sessionResult.getSession();
        assertNotNull(session, "Failed to create session");
    }

    @AfterEach
    void tearDown() {
        if (session != null) {
            try {
                session.delete();
            } catch (Exception e) {
                System.err.println("Warning: Failed to delete session: " + e.getMessage());
            }
        }
    }

    @Test
    void testSetBasicEnvVars() {
        Map<String, String> envs = new HashMap<>();
        envs.put("TEST_KEY", "test_value");
        envs.put("ANOTHER_KEY", "another_value");
        BoolResult result = session.getEnv().set(envs);
        assertTrue(result.isSuccess(), "set should succeed: " + result.getErrorMessage());
    }

    @Test
    void testGetAllEnvVars() {
        Map<String, String> envs = new HashMap<>();
        envs.put("SDK_TEST_A", "value_a");
        session.getEnv().set(envs);

        EnvResult result = session.getEnv().get();
        assertTrue(result.isSuccess());
        assertEquals("value_a", result.getEnvs().get("SDK_TEST_A"));
    }

    @Test
    void testGetSpecificKeys() {
        Map<String, String> envs = new HashMap<>();
        envs.put("GET_KEY_1", "val1");
        envs.put("GET_KEY_2", "val2");
        session.getEnv().set(envs);

        EnvResult result = session.getEnv().get("GET_KEY_1", "GET_KEY_2");
        assertTrue(result.isSuccess());
        assertEquals("val1", result.getEnvs().get("GET_KEY_1"));
        assertEquals("val2", result.getEnvs().get("GET_KEY_2"));
    }

    @Test
    void testOverwrite() {
        Map<String, String> envs = new HashMap<>();
        envs.put("OW_KEY", "original");
        session.getEnv().set(envs);
        assertEquals("original", session.getEnv().get("OW_KEY").getEnvs().get("OW_KEY"));

        envs.put("OW_KEY", "updated");
        session.getEnv().set(envs);
        assertEquals("updated", session.getEnv().get("OW_KEY").getEnvs().get("OW_KEY"));
    }

    @Test
    void testVisibleInShell() {
        Map<String, String> envs = new HashMap<>();
        envs.put("SHELL_VIS", "hello_from_env");
        session.getEnv().set(envs);

        CommandResult cmdResult = session.getCommand().executeCommand("echo $SHELL_VIS", 5000);
        assertTrue(cmdResult.isSuccess());
        assertTrue(cmdResult.getStdout().contains("hello_from_env"));
    }

    @Test
    void testSetEmptyThrows() {
        assertThrows(IllegalArgumentException.class, () -> {
            session.getEnv().set(new HashMap<>());
        });
    }
}
